lib/puppet/pops/types/type_factory.rb

```diff
@@ -351,6 +351,22 @@ def self.hash_of_data()
     type
   end
 
+  # Produces a type for NotUndef[T]
+  # The given 'inst_type' can be a string in which case it will be converted into
+  # the type String[inst_type].
+  #
+  # @param inst_type [Type,String] the type to qualify
+  # @return [Puppet::Pops::Types::PNotUndefType] the NotUndef type
+  #
+  # @api public
+  #
+  def self.not_undef(inst_type = nil)
+    type = Types::PNotUndefType.new()
+    inst_type = string(inst_type) if inst_type.instance_of?(String)
```

Comment:
is `kind_of?` also slow? (it may be just an alias).
My comment was that whenever you see `is_instance?` there is chance it is a typo and should be `is_a?`. ok to use `is_instance?`, but requires a comment so someone else does not change it, or has to figure out why it is used.